### PR TITLE
driver/azure: Performance improvements

### PR DESCRIPTION
**Pull request by ahmetalpbalkan**
Making Azure operation polling intervals more frequent to
detect completions faster (and thus making machine creations
faster).

Reduce one Azure API call for each machine command by caching
the discovered Azure Active Directory tenant ID of the user’s
Azure subscription into a local file. This also makes Azure
driver commands faster.

Removing the hacky polling around `azureutil.go#createStorageAccount`
by fixing it in Azure Go SDK.

Signed-off-by: Ahmet Alp Balkan <ahmetalpbalkan@gmail.com>

### CODE REVIEW

File: drivers/azure/azureutil/tenantid.go [hunk 1]
```diff
@@ -2,12 +2,55 @@ package azureutil
...
+func loadOrFindTenantID(env azure.Environment, subscriptionID string) (string, error) {
+	var tenantID string
+
+	// Load from cache
+	fp := tenantIDPath(subscriptionID)
+	b, err := ioutil.ReadFile(fp)
+	if err == nil {
+		tenantID = strings.TrimSpace(string(b))
+		log.Debugf("Tenant ID loaded from file: %s", fp)
+	} else {
```
Comment (dgageot) [inline drivers/azure/azureutil/tenantid.go]: Could this be an `else if`?

File: drivers/azure/azureutil/tenantid.go [hunk 2]
```diff
@@ -41,3 +84,7 @@ func findTenantID(env azure.Environment, subscriptionID string) (string, error)
 	}
 	return m[1], nil
 }
+
+func saveTenantID(path string, tenantID string) error {
+	return ioutil.WriteFile(path, []byte(tenantID), 0600)
```
Comment (dgageot) [inline drivers/azure/azureutil/tenantid.go]: Can we have multiple go routines writing at the same time the same file? Can this be problematic?

Comment (ahmetalpbalkan) [inline drivers/azure/azureutil/tenantid.go]: makes sense, I forgot about this as our `saveToken` handles this already. I'll turn this into write to tempfile and move to here.

File: drivers/azure/azureutil/tenantid.go (hunk 1 above)
Comment (ahmetalpbalkan) [inline drivers/azure/azureutil/tenantid.go]: fair point

File: drivers/azure/azureutil/tenantid.go (hunk 2 above)
Comment (dgageot) [inline drivers/azure/azureutil/tenantid.go]: Be careful, moving/renaming doesn't behave the same on linux/osx and windows.

**Comment by dgageot** [comment]
Other than the nitpicks, it LGTM!

Nice speedup. Creating an second azure machine:
Driver on master: 6m3
With this PR: 4m23

**Comment by ahmetalpbalkan** [comment]
@dgageot added another commit refactoring if/else and updating `saveTenantID` for atomic writes.

**Comment by ahmetalpbalkan** [comment]
@dgageot I've seen some 4m1s before. I know we can do better by potentially parallelizing creation of various independent resources required for the machines but this should be fair enough. :)

**Comment by ahmetalpbalkan** [comment]
@dgageot btw about your comment:

> Be careful, moving/renaming doesn't behave the same on linux/osx and windows.

what do you particularly refer to? We've been using it in storing azure tokens in docker-machine already and it worked fine on Windows and other OSes as far as I can tell.

**Comment by nathanleclaire** [comment]
Nice!  Looks fun.

File: drivers/azure/azureutil/clients.go [hunk 3]
```diff
@@ -49,6 +52,7 @@ func (a AzureClient) resourceGroupsClient() resources.GroupsClient {
 	c.Client.UserAgent += fmt.Sprintf(";docker-machine/%s", version.Version)
 	c.RequestInspector = withInspection()
```
Comment (ahmetalpbalkan) [inline drivers/azure/azureutil/clients.go]: @nathanleclaire  that could actually work, though as my note indicates, we're planning to make this stuff part of the SDK sometime soon via a factory pattern or something. So things will get better. I may actually submit a separate PR for that.

Comment (ahmetalpbalkan) [inline drivers/azure/azureutil/clients.go]: @nathanleclaire hmm I think the common base on all clients is `autorest.Client`, if we were to pass it as reference, that trick might do it, let's merge this and I'll take a stab at it next.

**Comment by nathanleclaire** [comment]
@ahmetalpbalkan Sure, sounds reasonable to address as a followup.

LGTM
